crates/llms/src/chat/mistral.rs

```diff
@@ -356,38 +386,44 @@ impl MistralLlama {
             .context(FailedToRunModelSnafu)?
             .send(self.to_mistralrs_request(
                 message.clone(),
-                false,
+                req.stream.unwrap_or_default(),
                 snd,
                 tools,
                 tool_choice,
-                sampling,
+                Some(sampling),
             ))
             .await
             .boxed()
             .context(FailedToRunModelSnafu)?;
         tracing::trace!("Request sent!");
 
-        match rcv.recv().await {
-            Some(response) => match response {
-                MistralResponse::Done(resp) => Ok(resp),
-                MistralResponse::ModelError(e, _) => {
-                    Err(ChatError::FailedToRunModel { source: e.into() })
-                }
-                MistralResponse::InternalError(e) | MistralResponse::ValidationError(e) => {
-                    tracing::error!(
-                        "Internal mistral.rs error: {e} for messages: {:#?}",
-                        message.clone()
-                    );
-                    Err(ChatError::FailedToRunModel { source: e })
-                }
-
-                // Don't expect MistralResponse::Chunk, should be streaming only.
-                _ => Err(ChatError::FailedToRunModel {
-                    source: "Unexpected error occurred".into(),
-                }),
-            },
-            None => Err(ChatError::FailedToRunModel {
-                source: "Mistral pipeline unexpectedly closed".into(),
+        Ok(rcv)
+    }
+
+    /// Process a single `stream=false` request to generate `OpenAi` chat completion.
+    fn post_process_req(resp: MistralResponse) -> Result<ChatCompletionResponse> {
+        match resp {
+            MistralResponse::Done(mut resp) => {
+                // mistralrs does not return "tool_calls" as a finish_reason correctly (like OpenAI spec).
+                // This is a workaround to set it correctly.
+                resp.choices.iter_mut().for_each(|c| {
+                    if c.finish_reason == "stop" && !c.message.tool_calls.is_empty() {
+                        c.finish_reason = "tool_calls".to_string();
+                    }
```

Comment:
Is the `finish_reason==stop` and `!tool_calls.is_empty()` sufficient to prove that the `finish_reason` is actually tool_call, could the finish reason something else (e.g. content filter) when both requirements are meet in this case.

Reply:
no, ` !tool_calls.is_empty(` is pretty definitive in my experience. There are other stop reasons for these cases `ContentFilter` or `Stop` if `max_completion_tokens` is reached.